event_substitution/wizard/wiz_event_substitution.py

```diff
@@ -21,36 +21,83 @@ class WizEventSubstitution(models.TransientModel):
+            vals = self._validate_employee_contract_and_registration(line)
+            self._validate_employee_calendar(
+                vals.get('date_start').year, vals.get('date_end').year,
+                line.employee.address_home_id)
+        threaded_calculation = threading.Thread(
+            target=self.button_substitution_employee_from_thread)
+        threaded_calculation.start()
+
+    def _validate_employee_contract_and_registration(self, line):
+        tz = self.env.user.tz
+        contract_obj = self.env['hr.contract']
+        vals = self._prepare_dates_for_registration(line)
+        date_from = _convert_to_local_date(
+            vals.get('date_start'), tz=tz).date()
+        date_to = _convert_to_local_date(
+            vals.get('date_end'), tz=tz).date()
+        cond = contract_obj._search_contracts_without_date_end(
+            line.employee.address_home_id, date_from)
+        permitted_contracts = contract_obj.search(cond)
+        cond = contract_obj._search_contracts_with_date_end(
+            line.employee.address_home_id, date_to, date_from)
+        permitted_contracts += contract_obj.search(cond)
+        if not permitted_contracts:
+            raise exceptions.Warning(
+                _("Employee %s without contract for event %s") %
+                (line.employee.name, line.event.name))
+        registrations = line.event.registration_ids.filtered(
+            lambda x: x.partner_id and x.date_start and
+            x.date_end and x.state in ('done', 'open') and
+            x.partner_id.id == line.employee.address_home_id.id and
+            ((str(vals.get('date_end')) >= x.date_start and
+              str(vals.get('date_end')) <= x.date_end) or
+             (str(vals.get('date_start')) <= x.date_end and
+              str(vals.get('date_start')) >= x.date_start)))
+        if registrations:
+            raise exceptions.Warning(
+                _('You can not confirm this registration, because their '
+                  'dates overlap with another record of the same employee,'
+                  ' in event: %s') % (line.event.name))
+        return vals
+
+    def _validate_employee_calendar(self, from_year, to_year, partner):
+        partner_calendar_obj = self.env['res.partner.calendar']
+        for year in range(from_year, to_year+1):
+            cond = [('partner', '=', partner.id),
+                    ('year', '=', year)]
+            if not partner_calendar_obj.search(cond):
+                raise exceptions.Warning(
+                    _('Employee %s without calendar') % (partner.name))
+
+    def button_substitution_employee_from_thread(self):
+        with openerp.api.Environment.manage():
+            with openerp.registry(self.env.cr.dbname).cursor() as new_cr:
+                new_env = api.Environment(
+                    new_cr, self.env.uid, self.env.context)
+                self.with_env(new_env).substitution_employee_from_thread()
+                new_env.cr.commit()
```

Comment:
¿No hay una manera mas sencilla de hacer esto? Según la documentación de Odoo si creas tu propio cursor haces commit, tienes que contemplar los casos en los que realizar el rollback.
¿Sabes que implicaciones puede tener esto? Lo digo por que yo no lo se.

https://doc.odoo.com/contribute/15_guidelines/coding_guidelines_framework/#never-commit-the-transaction

Reply:
Esto esta cogido de un ejemplo que me paso @mikelarre , y está cogido del core. En esta modificación del módulo, sólo se ha metido la creación del nuevo tree, para realizar un tratamiento que se ha realizado desde el principio de la aplicación. 
En caso de que hubiese un error dentro de este nuevo hilo, lo que implicaría es que no quedaría reflejada la sustitución del empleado que coge una baja, por el nuevo empleado, además la baja no quedaría confirmada, quedaría como pendiente de confirmación.